### PR TITLE
[mlir][tosa]: Add SLICE_SHAPE folder

### PR DESCRIPTION
**Pull request by udaya-ranga**
Add compile time folder for tosa.SLICE_SHAPE

### CODE REVIEW

File: mlir/test/Dialect/Tosa/constant_folding.mlir [hunk 1]
```diff
@@ -1469,4 +1469,16 @@ func.func @test_concat_shape_total_rank9_shapes() -> !tosa.shape<9> {
...
+
+// CHECK-LABEL: @test_slice_shape
+// CHECK: tosa.const_shape  {values = dense<[3, 4, 5, 6]> : tensor<4xindex>} : () -> !tosa.shape<4>
+func.func @test_slice_shape() -> !tosa.shape<4> {
+  %a = tosa.const_shape {values = dense<[1, 2, 3, 4, 5, 6]> : tensor<6xindex>} : () -> !tosa.shape<6>
+  %b = "tosa.const"() {values = dense<2> : tensor<1xi32>} : () -> tensor<1xi32>
+  %c = "tosa.const"() {values = dense<4> : tensor<1xi32>} : () -> tensor<1xi32>
+  %d = tosa.slice_shape %a, %b, %c : (!tosa.shape<6>, tensor<1xi32>, tensor<1xi32>) -> !tosa.shape<4>
+  return %d : !tosa.shape<4>
+}
```
Comment (lhutton1) [inline mlir/test/Dialect/Tosa/constant_folding.mlir]: nit: It would be nice to add some test cases for the negative cases e.g. no fold on start < 0, size <= 0, start + size > input size

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 2]
```diff
@@ -2093,6 +2103,52 @@ OpFoldResult concatShapeFold(tosa::ConcatShapeOp *op) {
...
+  if ((sizeV <= 0) || (startV < 0) ||
+      (static_cast<size_t>(startV + sizeV) > totalInput1))
+    return {};
+
+  SmallVector<APInt> sliceOfInput;
+  sliceOfInput.reserve(totalInput1);
+
+  for (auto i = startV; i < (startV + sizeV); i++) {
+    sliceOfInput.push_back(input1Vals[i]);
+  }
```
Comment (lhutton1) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp]: nit: could be simplified to?
```
SmallVector<APInt> sliceOfInput(
    input1Vals.begin() + startV,
    input1Vals.begin() + startV + sizeV
);
```

**Review by lhutton1 — COMMENTED** [review]
Had a couple of small comments, otherwise LG

**Review by lhutton1 — APPROVED** [review]
~Will address the nitpicks in a follow-up shortly~ see above
